### ISSUE TITLE
CloudFormation: Assign elastic IPs from a certain pool to new nodes 

### ISSUE THREAD

**Issue by Thermi** [coreos/kube-aws#219]
Hello,

I'd like to see the feature in the CloudFormation configuration that assigns elastic IPs from a certain pool of addresses (just a list of EIPs) to the new nodes, when they are created by CloudFormation.

My use case is that the cluster has to be reachable from a certain set of IPs, as only the traffic to those addresses is zero rated for the clients. That means that the clients do not have to pay for the traffic. Or the traffic is not added to their data usage.

It would be quite handy to have this feature, because currently, when nodes are rotated, I need to manually assign the EIPs again.

**Comment by redbaron** [comment]
as nodes are created in AutoScaling groups there is no straightforward way to assign EIPs to each one of them.

What you can do is to modify `stack-template.json` and create ENI interfaces with EIP assigned to them, tag them accordingly, then have a script on nodes where you want to receive traffic to call AWS to attach additional network interface

**Comment by mumoshu** [comment]
@Thermi Thanks for your interest on kube-aws.

In addition to what @redbaron described, I'd also recommend taking a look into @neoandroid's great work #169 which allows you to attach worker nodes a NAT gateway with your elastic IP allocated on it.

**Comment by mumoshu** [comment]
@Thermi Would you mind confirming if any of what @redbaron and I provided (will) resolve your issue?
// I'd generally recommend NAT gateways with elastic IPs for your use-case for better availability

**Comment by Thermi** [comment]
@mumoshu We require each node to have its own special elastic IP, so we can't use @neoandroid's work. I think what we'll end up doing is using preallocated network interfaces. How does kubernetes and coreos behave when a network interface is attached when the node is already running? In our case, we'd have to switch over all connections to the new connection by adding corresponding routes into a special routing table and routing rules. So this is very complex.

**Comment by mumoshu** [comment]
@Thermi Thanks for your reply!

Could you elaborate a bit further?
I guess putting worker nodes in the back of a NAT gateway with an EIP will make all the packets coming from worker nodes as the ones from the specific EIP. So it will be zero-rated according to your original explanation, even there're multiple backend worker nodes in a private subnet(you gain HA and zero-rated transfer without complex code!).

Also see #284 for my recent work to achieve worker nodes in a private subnet while having a NAT gateway with a preallocated EIP assigned in a public subnet, which I believe result in achieving zero-rated H/A worker nodes in your case.

**Comment by mumoshu** [comment]
Hi @Thermi, did you have a chance to read my comment above? 😃

**Comment by Thermi** [comment]
@mumoshu Well, yes, if we used a NAT gateway, that would work, but it would
1) restrict the available bandwidth and prevent a reuse of any port globally
2) not give us HA, because a single instance (which a NAT gateway is, as far as I know) is not HA
3) we couldn't reach any of the instances via SSH, except by connecting to the NAT gateway first.

**Comment by neoandroid** [comment]
1. NAT gateway support up to 10Gbps. [1]  Do you need more than that? If that's the case then you'd need a more complex setup to achieve bigger throughput.
2. NAT gateways are implemented w/ redundancy in each zone, so they are HA. [1] The only way to break HA would be to have EC2 instances spread across all AZ and having NAT gateways only in one of the AZ.
3. You're right, you'd have to setup your own bastion host. NAT gateways are not instances  we're you can  connect to via SSH, they're managed by Amazon.

[1] [AWS Documentation](http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/vpc-nat-gateway.html)

**Comment by Thermi** [comment]
Well, if this endeavour goes well, the bandwidth will be very high, so I don't want to lock into the NAT gateway solution.
The point about port reuse still stands.

**Comment by Thermi** [comment]
How does CoreOS behave when there are multiple network interfaces to a VPC?

**Comment by mumoshu** [comment]
@Thermi I think I now got your point on sticking to EIPs directly assigned to nodes.
However I'm afraid I can't personally help you soon as I'm not a CoreOS guy who knows a detail like that!
I suggest you to consult someone more experienced to CoreOS about your last question https://github.com/coreos/kube-aws/issues/219#issuecomment-277275044

**Comment by mumoshu** [comment]
Btw @Thermi 
> It would be quite handy to have this feature, because currently, when nodes are rotated, I need to manually assign the EIPs again.

Just curious but isn't it possible to assign an EIP by writing your own custom bash script which just automate what you're doing manually today, in `userdata/cloud-config-*`?
That way you can virtually hook into the startup sequence of your nodes to assign EIPs automatically.

**Comment by Thermi** [comment]
@mumoshu No, just doing that wouldn't cut it, because every instance has the same cloud config. We'd need to do some state keeping on another host or in other infrastructure to make sure we don't set the same EIP that is already in use or do other bad stuff. But the idea is good, I just wouldn't do it with Bash, but in another language.

**Comment by redbaron** [comment]
@Thermi you can use opportunistic approach: every node queries unassigned EIPs and tries to assign randomly chosen one. if failed (which means that somebody else grabbed it first), then it fetches new list of unassigned EIPs and tries again.

**Comment by mumoshu** [comment]
@Thermi FYI, inspired by our discussion here, I've utilized the pool-of-EIPs technic in #332. 

Regarding the state keeping you've mentioned, I've used a pool of pairs of an EBS volume and an EIP.

More concretely, in #332, every etcd node tries to attach an EBS volume from the pool via cloud-config at startup. If the volume attachment was unsuccessful, repeat until it succeeds. Once a volume is successfully attached, determine a paired EIP via a tag named `kube-aws:etcd:eip-allocation-id` added to the EBS volume, and then associate it to the node. This way, I managed to achieve the pool of EIPs without a serious database to keep states.

It won't be always applicable but just fyi.

**Comment by Thermi** [comment]
@mumoshu What's up with the "I couldn't make it work"?

**Comment by mumoshu** [comment]
@Thermi I couldn't come up with the theoretically minimum set of IAM policies that work for my use-case. I don't remember the exact sets of policies I've tried but none of them worked in my experiments.

**Comment by mumoshu** [comment]
Hi @Thermi - FYI we now have `customSystemdUnits` to inject any systemd units as you like.
For example, this can be used to assign an EIP chosen from a pool to a worker node.
